x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/waitforallocations/ScalingInference.java

```diff
@@ -0,0 +1,138 @@
+
+    private final Map<String, LinkedBlockingQueue<WaitingRequest>> queueRequests;
+
+    public record WaitingRequest(
+        InferModelAction.Request request,
+        InferModelAction.Response.Builder responseBuilder,
+        TaskId parentTaskId,
+        ActionListener<InferModelAction.Response> listener
+    ) {
+        public String deploymentId() {
+            return request.getId();
+        }
+    }
+
+    private final TrainedModelAssignmentService assignmentService;
+    private final BiConsumer<WaitingRequest, TrainedModelAssignment> queuedConsumer;
+
+    public ScalingInference(
+        TrainedModelAssignmentService assignmentService,
+        BiConsumer<WaitingRequest, TrainedModelAssignment> queuedConsumer
+    ) {
+        this.assignmentService = assignmentService;
+        this.queuedConsumer = queuedConsumer;
+        this.queueRequests = new ConcurrentHashMap<>();
+    }
+
+    public synchronized void waitForAssignment(WaitingRequest request) {
+        var p = queueRequests.computeIfAbsent(request.deploymentId(), k -> new LinkedBlockingQueue<>());
+
+        if (p.isEmpty()) {
+            logger.info("waitForAssignment will wait for condition");
+            p.offer(request);
+            assignmentService.waitForAssignmentCondition(
+                request.deploymentId(),
+                new DeploymentHasAtLeastOneAllocation(request.deploymentId()),
+                request.request().getInferenceTimeout(),
+                new WaitingListener(request.deploymentId())
+            );
+        } else {
+            logger.info("waitForAssignment added to queue");
+            p.offer(request);
+        }
+
+    }
+
+    private static class DeploymentHasAtLeastOneAllocation implements Predicate<ClusterState> {
+
+        private final String deploymentId;
+        private Exception exception;
+
+        DeploymentHasAtLeastOneAllocation(String deploymentId) {
+            this.deploymentId = ExceptionsHelper.requireNonNull(deploymentId, "deployment_id");
+        }
+
+        @Override
+        public boolean test(ClusterState clusterState) {
+            logger.info("predicate test");
+            TrainedModelAssignment trainedModelAssignment = TrainedModelAssignmentMetadata.assignmentForDeploymentId(
+                clusterState,
+                deploymentId
+            ).orElse(null);
+            if (trainedModelAssignment == null) {
+                logger.info(() -> format("[%s] assignment was null while waiting to scale up", deploymentId));
+                return false;
+            }
+
+            var allocationStatus = trainedModelAssignment.calculateAllocationStatus().orElse(null);
+            if (allocationStatus == null) {
+                return false;
+            }
+
+            var state = allocationStatus.calculateState();
+            logger.info("calculated state " + state);
+
+            return state.isAnyOf(AllocationStatus.State.STARTED, AllocationStatus.State.FULLY_ALLOCATED);
```

Comment:
If a deployment has `target_allocation_count` of zero, it's consider fully allocated. Is this logic correct in that case?

For `DeploymentHasAtLeastOneAllocation` I'd expect a condition `num_allocations >= 1`.

Reply:
👍 good catch